src/storage/src/source/mysql/replication.rs

```diff
@@ -313,50 +315,71 @@ pub(crate) fn render<G: Scope<Timestamp = GtidPartition>>(
-                                return Ok(return_definite_error(
-                                    err,
-                                    &output_indexes,
-                                    &mut data_output,
-                                    data_cap_set,
-                                    &mut definite_error_handle,
-                                    definite_error_cap_set,
-                                )
-                                .await);
-                            }
-                            let new_upper = repl_partitions.frontier();
-                            repl_context.advance("xid_event", new_upper);
+                        // Increment the transaction-id to the next GTID we should see from this source-id
+                        let next_tx_id = tx_id.checked_add(1).unwrap();
+                        let next_gtid =
+                            GtidPartition::new_singleton(source_id, GtidState::Active(next_tx_id));
+
+                        if let Err(err) = data_partitions.try_downgrade(next_gtid) {
+                            return Ok(return_definite_error(
+                                err,
+                                &output_indexes,
+                                &mut data_output,
+                                data_cap_set,
+                                &mut definite_error_handle,
+                                definite_error_cap_set,
+                            )
+                            .await);
                         }
+                        let new_upper = data_partitions.frontier();
+                        repl_context.downgrade_data_cap_set("xid_event", new_upper);
                     }
                     // We receive a GtidEvent that tells us the GTID of the incoming RowsEvents (and other events)
                     Some(EventData::GtidEvent(event)) => {
                         let source_id = Uuid::from_bytes(event.sid());
-                        let received_tx_id =
-                            GtidState::Active(NonZeroU64::new(event.gno()).unwrap());
+                        let tx_id = NonZeroU64::new(event.gno()).unwrap();
+
+                        // We are potentially about to ingest a big transaction that we don't want
+                        // to store in memory. For this reason we are immediately downgrading our
+                        // progress frontier to one that includes the upcoming transaction. This
+                        // will cause a remap binding to be minted right away and so the data of
+                        // the transaction will not accumulate in the reclock operator.
+                        let next_tx_id = tx_id.checked_add(1).unwrap();
+                        let next_gtid =
+                            GtidPartition::new_singleton(source_id, GtidState::Active(next_tx_id));
+
+                        if let Err(err) = progress_partitions.try_downgrade(next_gtid) {
```

Comment:
nit: the `try_downgrade` name is slightly confusing since we also have the `downgrade_<>_cap_set()` method called below (they sound like they do the same thing). Would `progress_partitions.advance_frontier(next_gtid)` make more sense?

Reply:
changed to `advance_frontier`. My hope is that we'll get rid of these structures once it's safe to keep around capabilities in async operators